compiler/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -909,6 +953,142 @@ impl<'interner> Monomorphizer<'interner> {
+    /// structs to positions in the runtime tuple, since all structs are
+    /// replaced by tuples during compilation.
+    fn patch_debug_member_assign(
+        &mut self,
+        call: &HirCallExpression,
+        arguments: &mut [ast::Expression],
+        arity: usize,
+    ) {
+        let hir_arguments = vecmap(&call.arguments, |id| self.interner.expression(id));
+        let Some(HirExpression::Literal(HirLiteral::Integer(fe_var_id, _))) = hir_arguments.get(0) else {
+            unreachable!("Missing FE var ID in __debug_member_assign call");
+        };
+        let Some(HirExpression::Ident(HirIdent { id, .. })) = hir_arguments.get(1) else {
+            unreachable!("Missing value identifier in __debug_member_assign call");
+        };
+        // update variable member assignments
+        let var_def_name = self.interner.definition(*id).name.clone();
+        let var_type = self.interner.id_type(call.arguments[2]);
+        let fe_var_id = fe_var_id.to_u128() as u32;
+
+        let mut cursor_type = self
+            .debug_types
+            .get_type(fe_var_id)
+            .unwrap_or_else(|| panic!("type not found for fe_var_id={fe_var_id}"))
+            .clone();
+        for i in 0..arity {
+            if let Some(HirExpression::Literal(HirLiteral::Integer(fe_i, i_neg))) =
+                hir_arguments.get(2 + i)
+            {
+                let mut index = fe_i.to_i128();
+                if *i_neg {
+                    index = -index;
+                }
+                if index < 0 {
+                    let index = index.unsigned_abs();
+                    let field_name = self
+                        .debug_field_names
+                        .get(&(index as u32))
+                        .unwrap_or_else(|| panic!("field name not available for {i:?}"));
+                    let field_i = (get_field(&cursor_type, field_name)
+                        .unwrap_or_else(|| panic!("failed to find field_name: {field_name}"))
+                        as i128)
+                        .unsigned_abs();
+                    cursor_type = next_type(&cursor_type, field_i as usize);
+                    let index_id = self.interner.push_expr(HirExpression::Literal(
+                        HirLiteral::Integer(field_i.into(), false),
+                    ));
+                    self.interner.push_expr_type(&index_id, Type::FieldElement);
+                    self.interner.push_expr_location(
+                        index_id,
+                        call.location.span,
+                        call.location.file,
+                    );
+                    arguments[2 + i] = self.expr(index_id);
+                } else {
+                    cursor_type = next_type(&cursor_type, 0);
+                }
+            } else {
+                cursor_type = next_type(&cursor_type, 0);
+            }
+        }
+
+        let var_id = if &var_def_name != "__debug_expr" {
```

Comment:
This check seems a bit fragile

Reply:
Agreed, and this actually shouldn't be necessary because when doing member assignment, the variable should be already instantiated for the current type and we will always assign from `__debug_expr`. I'm thinking this logic is a leftover from a previous implementation. In fact, we should be able to capture the names of the variables during injection and do the instantiation/lookup by type here without needing the name.

Reply:
This bit of logic changed in the last commits. The name of the variables are now carried over from the instrumentation phase, so there's no need to check of the value passed into the oracle is the variable itself or the temporary variable `__debug_expr`. This was brittle as you said.